module/VuFind/src/VuFindTest/Feature/AutoRetryTrait.php

```diff
@@ -107,6 +107,13 @@ public function runBare(): void
                 // Execute callbacks for interrupted test, unless this is the
                 // last round of testing:
                 if ($this->retriesLeft > 0) {
+                    $this->logWarning(
```

Comment:
The AutoRetryTrait is not _necessarily_ always going to be used by Mink test cases (there might be other types of integration tests that could benefit from this), so I don't think it's safe to assume that the logWarning method is defined in this context.

A few possible solutions:

1.) We could implement a new `@retryLogMethod` annotation to tell the trait what to call (if anything).

2.) We could define `logWarning` as part of AutoRetryTrait (or give it a separate trait that AutoRetryTrait then uses in turn).

3.) We could check to see if the method exists before calling it.

I tend to lean toward option 1, though that might be annoying since it would require us to annotate a whole bunch of classes, and we could easily forget that detail. Maybe a combination of all three ideas would be most flexible: allow an annotation to override the callback function, but default to logWarning, and check that the method exists before calling it.

(I think the AutoRetryTrait is a useful piece of code that others might find valuable outside of VuFind's context, so I want to be mindful about keeping it as flexible/reusable as we can).

Reply:
Combination added.